src/bigquery.ts

```diff
@@ -2391,9 +2371,228 @@ promisifyAll(BigQuery, {
+  schemaField: TableField,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  value: any,
+  options: {
+    wrapIntegers: boolean | IntegerTypeCastOptions;
+    selectedFields?: string[];
+    parseJSON?: boolean;
+  }
+) {
+  if (is.null(value)) {
+    return value;
+  }
+
+  switch (schemaField.type) {
+    case 'BOOLEAN':
+    case 'BOOL': {
+      value = value.toLowerCase() === 'true';
+      break;
+    }
+    case 'BYTES': {
+      value = Buffer.from(value, 'base64');
+      break;
+    }
+    case 'FLOAT':
+    case 'FLOAT64': {
+      value = Number(value);
+      break;
+    }
+    case 'INTEGER':
+    case 'INT64': {
+      const {wrapIntegers} = options;
+      value = wrapIntegers
+        ? typeof wrapIntegers === 'object'
+          ? BigQuery.int(
+              {integerValue: value, schemaFieldName: schemaField.name},
+              wrapIntegers
+            ).valueOf()
+          : BigQuery.int(value)
+        : Number(value);
+      break;
+    }
+    case 'NUMERIC': {
+      value = new Big(value);
+      break;
+    }
+    case 'BIGNUMERIC': {
+      value = new Big(value);
+      break;
+    }
+    case 'RECORD': {
+      value = BigQuery.mergeSchemaWithRows_(schemaField, value, options).pop();
+      break;
+    }
+    case 'DATE': {
+      value = BigQuery.date(value);
+      break;
+    }
+    case 'DATETIME': {
+      value = BigQuery.datetime(value);
+      break;
+    }
+    case 'TIME': {
+      value = BigQuery.time(value);
+      break;
+    }
+    case 'TIMESTAMP': {
+      const pd = new PreciseDate();
+      pd.setFullTime(PreciseDate.parseFull(BigInt(value) * BigInt(1000)));
+      value = BigQuery.timestamp(pd);
+      break;
+    }
+    case 'GEOGRAPHY': {
+      value = BigQuery.geography(value);
+      break;
+    }
+    case 'JSON': {
+      const {parseJSON} = options;
+      value = parseJSON ? JSON.parse(value) : value;
+      break;
+    }
+    case 'RANGE': {
+      value = BigQueryRange.fromSchemaValue_(
+        value,
+        schemaField.rangeElementType!.type!
+      );
+      break;
+    }
+    default:
+      break;
+  }
+
+  return value;
+}
+
+/**
+ * Range class for BigQuery.
+ * A range represents contiguous range between two dates, datetimes, or timestamps.
+ * The lower and upper bound for the range are optional.
+ * The lower bound is inclusive and the upper bound is exclusive.
+ * See https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#range_literals
+ */
+export class BigQueryRange {
+  elementType?: string;
+  start?: BigQueryTimestamp | BigQueryDate | BigQueryDatetime;
+  end?: BigQueryTimestamp | BigQueryDate | BigQueryDatetime;
+  constructor(value: string | BigQueryRangeOptions, elementType?: string) {
+    if (typeof value === 'string') {
+      if (!elementType) {
+        throw new Error(
+          'invalid RANGE. Element type required when using RANGE literal string.'
+        );
+      }
+      let cleanedValue = value;
```

Comment:
This section seems to be the same code as `static fromSchemaValue_()`- does it make sense to reuse code here?

Reply:
good tip, I moved that logic to a `private fromStringValue_` method